traits/src/grid.rs

```diff
@@ -21,21 +23,36 @@ pub trait Topology {
     fn dim(&self) -> usize;
 
     // Check the locality of an element
-    fn locality(global_id: usize) -> Locality;
+    fn locality(&self, global_id: usize) -> Locality;
 
     // Convert local to global id
-    fn local2global(local_id: usize) -> usize;
+    fn local2global(&self, local_id: usize) -> usize;
 
     // Convert global to local id
-    fn global2local(global_id: usize) -> Option<usize>;
+    fn global2local(&self, global_id: usize) -> Option<usize>;
+
+    fn entity_count(&self, dim: usize) -> usize;
+
+    fn cell(&self, index: usize) -> Option<&[usize]>;
+    unsafe fn cell_unchecked(&self, index: usize) -> &[usize];
+
+    fn create_connectivity(&mut self, dim0: usize, dim1: usize);
+    fn create_connectivity_all(&mut self) {
+        for dim0 in 0..self.dim() {
+            for dim1 in 0..self.dim() {
+                self.create_connectivity(dim0, dim1);
+            }
+        }
+    }
+    fn connectivity(&self, dim0: usize, dim1: usize) -> &AdjacencyList<usize>;
 }
 
 pub trait Grid {
-    type Topology<'a>: Topology
-    where
-        Self: 'a;
+    type Topology: Topology;
+    type Geometry: Geometry;
 
-    fn topology<'a>(&'a self) -> Self::Topology<'a>;
+    fn topology(&self) -> &Self::Topology;
+    fn topology_mut(&mut self) -> &mut Self::Topology;
 
-    fn cell_geometry<'a>(&'a self, local_index: usize) -> &'a dyn Geometry;
```

Comment:
What is the difference between cell_geometry and geometry? If there is a difference should we have better names?

Reply:
The geometry contains the geometric information about where points are. I've removed cell_geometry for now. A cell will need to know about both its geometry and its topology, so cell_geometry in any case was a bad name